### PR TITLE
Recon x64

### PR DESCRIPTION
**Pull request by cmaruti**
Improved support for x64 systems

### CODE REVIEW

File: vdb/recon/__init__.py [hunk 1]
```diff
@@ -68,6 +68,21 @@ def reprargs(trace, fmt, args):
...
     return r
 
+def getArgs(trace, args):
+    '''
+    Assuming we are at the instruction after
+    a call, grab the argument at the specified
+    index (skipping the saved instruction pointer).
+    '''
+    cc = trace.getEmulator().getCallingConvention("stdcall")
+    if "64" in trace.getMeta("Architecture"):
```
Comment (invisig0th) [inline vdb/recon/__init__.py]: Love the overall idea, however for future proofing, we cannot assume that any arch with the chars "64" in the name ( such as commodore64 ) are amd64.  This probably needs to be come something more like a dict lookup of
(arch,plat):cc

File: vdb/recon/__init__.py [hunk 2]
```diff
@@ -68,6 +68,21 @@ def reprargs(trace, fmt, args):
...
+        if trace.getMeta("Platform") == "windows":
+            cc = trace.getEmulator().getCallingConvention("msx64call")
+        else: trace.getEmulator().getCallingConvention("sysvamd64call")
```
Comment (invisig0th) [inline vdb/recon/__init__.py]: inline "else: foo" is not style compatible with the rest of the code, i know it's a nit but if we dont try to "dig up" the code base will only become more inconsistent :)

File: vdb/recon/__init__.py [hunk 3]
```diff
@@ -85,8 +100,8 @@ def notify(self, event, trace):
         stackptr = trace.getStackCounter()
 
         rawargs = trace.readMemoryFormat(stackptr, '<%dP' % (len(self._reconfmt) + 1))
-        savedeip = rawargs[0]
-        args = rawargs[1:]
+        savedeip = trace.getProgramCounter()
+        args = getArgs(trace, len(self._reconfmt))
```
Comment (invisig0th) [inline vdb/recon/__init__.py]: perfect.  heading down this road, we'll hopefully manage to support windows arm without too much trouble :)

File: vdb/recon/sniper.py [hunk 4]
```diff
@@ -5,16 +5,24 @@
...
     '''
     Assuming we are at the instruction after
-    a call, grab the stack argument at the specified
+    a call, grab the argument at the specified
     index (skipping the saved instruction pointer).
     '''
-    stack = trace.getStackCounter()
+    cc = trace.getEmulator().getCallingConvention("stdcall")
+
+    if "64" in trace.getMeta("Architecture"):
```
Comment (invisig0th) [inline vdb/recon/sniper.py]: another instance of "fuzzy" substring comparison.  on amd64, the "Architecture" meta will always be exactly "amd64" so lets test for that.

**Comment by cmaruti** [comment]
Do you need something else here man before merge?

**Comment by atlas0fd00m** [comment]
@cmaruti does this PR still function as expected on current Viv/Vdb?

**Comment by cmaruti** [comment]
I don’t know atm; gonna check it

**Comment by cmaruti** [comment]
@atlas0fd00m seems the whole breakpoint stuffs on win32 (32/64) does not work anymore :(

BEFORE merging master branch into
>  python sample06_sniper.py
RECON: 1260 0x76c3ec44 user32.CreateWindowExW(256, u'CalcFrame', u'Calcolatrice', 13238272)
RECON: 1260 0x76c3ec44 user32.CreateWindowExW(0, u'CalcFrame', NULL, 1073741824)
RECON: 1260 0x76c3ec44 user32.CreateWindowExW(0, u'EDIT', u'', 268435456)
RECON: 1260 0x76c3ec44 user32.CreateWindowExW(0, u'STATIC', u'', 1342177358)
RECON: 1260 0x76c3ec44 user32.CreateWindowExW(0, 0x0000c03b, u'OleMainThreadWndName', 2281701376L)
RECON: 1260 0x76c3ec44 user32.CreateWindowExW(0, u'CicMarshalWndClass', u'CicMarshalWnd', 2281701376L)
RECON: 1260 0x76c3ec44 user32.CreateWindowExW(0, u'MSCTFIME UI', u'MSCTFIME UI', 2281701376L)

AFTER merging master branch
> NO breakpoints hit

```python
import sys
VDB_ROOT = "C:\\vivisect"
sys.path.append(VDB_ROOT)

import vtrace
import vtrace.breakpoints as v_bp
import vdb.recon as v_rec
import vdb.recon.sniper as v_snip

def hookfunc(event, trace, retaddr, args, cc):
    print('ret: 0x%.8x: %r' % (retaddr, args))
    
trace = vtrace.getTrace()

#v_snip.snipeDynArg(trace, 'user32.CreateWindowExW', 3)
#v_snip.snipeArgValue(trace, 'user32.CreateWindowExW', 1, 0x100)
#v_bp.addHook( trace, 'user32.CreateWindowExW', hookfunc )
v_rec.addReconBreak(trace, 'user32.CreateWindowExW', 'IUUI')

trace.execute(r'c:\windows\system32\calc.exe')
trace.setMode('RunForever', True)
trace.run()

**Comment by atlas0fd00m** [comment]
Windows has added significant changes to the debugging security model, requiring more permissions to *write* to a debugged process.  since Breakpoints actually overwrite bytes in the target process (for Intel, Viv writes the Int3 instruction, or 0xCC at the breakpoint location).

but wait, you're suggesting that this change has to do with an update to Viv?

i apologize for the delay in responding, @cmaruti .  i didn't realize you had (bad notification mechanisms for me apparently).

can you give me more information about what version you were using before breakpoints stopped working?  a date?

**Comment by cmaruti** [comment]
Since a long time has passed can't be helpful. Probably the best approx was the date around the post before my last one

**Comment by atlas0fd00m** [comment]
my apologies, @cmaruti , apparently github stopped sending me email notifications and i only now realized it.
thank you for your contributions.

i'm working out why this stopped working for you and will have a PR done soon.

**Comment by cmaruti** [comment]
Np sir! I'll be more than happy to check it out later when done

**Comment by atlas0fd00m** [comment]
hey @cmaruti 
thank you for your patience...
could you see if this branch of Viv changes anything for you?

i'm not sure if this is a necessary change, since i think i've been running into issues where the user account simply lacks the permissions overall.  but that doesn't answer *your* problem.

https://github.com/atlas0fd00m/vivisect/tree/win_vdb_privs

https://github.com/vivisect/vivisect/pull/536


thanks!
@

**Comment by cmaruti** [comment]
Hi @atlas0fd00m nothing changed from my side. Now I got the following trace and seems problem is elsehwere...

```
C:\vivisect-win_vdb_privs>python sample.py
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
Failed to resolve breakpoint address for expression: user32.CreateWindowExW
Error:
Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 22, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 37, in evaluate
    val = eval(pycode, {}, locvars)
  File "<string>", line 1, in <module>
NameError: name 'user32' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\breakpoints.py", line 98, in resolveAddress
    self.address = trace.parseExpression(self.vte)
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\vtrace\__init__.py", line 939, in parseExpression
    return int(e_expr.evaluate(expression, vlocs))
  File "C:\vivisect-win_vdb_privs\vivisect-win_vdb_privs\envi\expression.py", line 40, in evaluate
    raise ExpressionFail(pycode, e)
envi.expression.ExpressionFail: ExpressionFail: 'user32.CreateWindowExW' is not a valid expression in this context (NameError("name 'user32' is not defined"))
```

**Comment by atlas0fd00m** [comment]
thanks for checking.
that is certainly odd.  a very unexpected error.
that looks like it's failing to lookup expressions, looking for user32 functions.

can you give me any more specifics about what you think is causing this?  i don't recognize that issue so it's hard to track it down on my side.  

also, did you install pywin32?

**Comment by atlas0fd00m** [comment]
can you share what's in `sample.py`?

**Comment by cmaruti** [comment]
Yep was totally unexpected also for me... 
pywin32 was installed and here you go with the contents of  `sample.py`

```
import sys
VDB_ROOT = "C:\\Users\\cmaruti\\Downloads\\vivisect-win_vdb_privs\\vivisect-win_vdb_privs\\"
sys.path.append(VDB_ROOT)

import vtrace
import vtrace.breakpoints as v_bp
import vdb.recon as v_rec
import vdb.recon.sniper as v_snip

def hookfunc(event, trace, retaddr, args, cc):
    print('ret: 0x%.8x: %r' % (retaddr, args))
    
trace = vtrace.getTrace()

#v_snip.snipeDynArg(trace, 'user32.CreateWindowExW', 3)
#v_snip.snipeArgValue(trace, 'user32.CreateWindowExW', 1, 0x100)
#v_bp.addHook( trace, 'user32.CreateWindowExW', hookfunc )
v_rec.addReconBreak(trace, 'user32.CreateWindowExW', 'IUUI')

trace.execute(r'c:\windows\system32\calc.exe')
trace.setMode('RunForever', True)
trace.run()
```

**Comment by atlas0fd00m** [comment]
thanks, @cmaruti . i'll look into it.  

just to level-set on your environment, could you try using VDB on your machine, both mainline and this branch and let me know if you are able to set breakpoints on both (just standard VDB).

thanks@

**Comment by cmaruti** [comment]
Hi @atlas0fd00m and sorry for the late reply; I have been away from the pc for a while

I did a silly mistake in my script tracked down thanks to your input :D! I can confirm both of the branches (main and priv) work well with the recon commit on my Win10 box; I have also update the posted script for your convenience

File: vdb/recon/__init__.py [hunk 5]
```diff
@@ -68,18 +68,28 @@ def reprargs(trace, fmt, args):
         r.append(rstr)
     return r
 
+def detect_cc(trace):
```
Comment (atlas0fd00m) [inline vdb/recon/__init__.py:72]: i like this approach (calling `detect_cc`) much better than inline hand-jamming...

please define the dictionary outside the function, so it doesn't get recreated every call.

also, these are the defaults (although i386/linux should be cdecl), but not necessarily the reality.  what happens when the code gets it wrong?  what's worst-case scenario?

thanks!

File: vdb/recon/sniper.py [hunk 6]
```diff
@@ -11,18 +11,24 @@ def getArg(trace, argidx):
...
-        else: cc = trace.getEmulator().getCallingConvention("sysvamd64call")
-    
+    cc = detect_cc(trace)    
     args = cc.getCallArgs(trace, argidx)
     return args[-1]
-    '''stack = trace.getStackCounter()
-    fmt = '<P' + ('P' * (argidx+1))
-    args = trace.readMemoryFormat(stack, fmt)
-    return args[-1]'''
+
```
Comment (atlas0fd00m) [inline vdb/recon/sniper.py:17]: please only one copy of `detect_cc`

**Comment by cmaruti** [comment]
I did my homework; hope it is fine @atlas0fd00m :D

**Comment by rakuy0** [comment]
@cmaruti Neat. Looks good so far.

You mind throwing in a unit test kinda demonstrating it in action (and so @atlas0fd00m/myself know if we accidentally break it when we're down in vdb town).

**Comment by atlas0fd00m** [comment]
hey @cmaruti ,  looking good, thanks for the improvements.  

let me know if you want to chat about some good unittests (and how to wrap them into our testing framework).  that's all that's keeping us from merging this in.  well, that and merging in the base branch.

@rakuy0 and i just want to be sure we know if we break something moving forward.  keeps us honest and improves the overall quality of vivisect.  if you look back through our merged PR's, he and i are pretty frequently saying "hey, looks good!  where's the tests?"  lol.

thanks, man,
@

**Comment by cmaruti** [comment]
Sorry guys but unfortunately I had no time to put togheter an unit test but I will do in the upcoming weeks. I'm not good at that but I'll try to get by

File: vdb/tests/testrecon.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    '''
+    Tests to verify recon hooking during execution.
+    '''
+
+    @classmethod
+    def setUpClass(cls):
+        cls.hello_i386_vw = helpers.getTestWorkspace('windows', 'i386',
+                                                     'helloworld.exe')
+        cls.hello_amd64_vw = helpers.getTestWorkspace('windows', 'amd64',
+                                                      'helloworld.exe')
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vdb/tests/testrecon.py:26]: alternately, making it support all the Intel platforms would be even better.

**Review by atlas0fd00m — CHANGES_REQUESTED** [review]
sorry for the delay in response.  i've been sitting on this review, trying to get the libc.exit hook fixed.  i now have a fix in, and ready for moving forword.

currently your unittest fails.  let's figure out why and then we can get this merged in.
unfortunately i believe part of this failure could be waiting on #550

File: vdb/tests/testrecon.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    def test_vdb_recon_i386(self):    
+        trace = vtrace.getTrace()
+        plat = trace.getMeta('Platform')
+        sym = sym_name.get(plat)
+		
+        trace.execute(helpers.getTestPath('windows', 'i386',
+                                                     'helloworld.exe'))
```
[the remark this answers is not in the document]
Comment (cmaruti) [reply vdb/tests/testrecon.py:34]: I am going to approve all of your suggested changes; in the meantime I'll try to find a sample in the vivtestfile repo that fit the use case and later extend the unit test

**Comment by atlas0fd00m** [comment]
@cmaruti , could you pull in the latest master branch?  @invisig0th just merged in a fix allowing Linux systems to pick-up library loads as they occur (which would affect your unittests).

thanks!
@

**Comment by cmaruti** [comment]
@atlas0fd00m CI tests failed once again :(

**Comment by atlas0fd00m** [comment]
so the other error went away tho.  recon *is* working on AMD64.  i'm guessing it is failing on i386 because of a test issue rather than a recon/vdb issue.  we're getting close!

**Comment by atlas0fd00m** [comment]
@cmaruti , would you like any help on this?
DM me in Synapse Slack if you'd like more back-and-forth conversation.

thanks again!
@

**Comment by cmaruti** [comment]
Hi @atlas0fd00m and sorry for the late response...
I tracked down the problem and I'm pretty sure the unittest fails because circle-ci wasn't able to run 32-bit binary. I have the same behavior on my box; after I have add support for 32-bits elf (sudo dpkg --add-architecture i386; sudo apt-get install libc6:i386 libncurses5:i386 libstdc++6:i386) unit tests complete succesfully. 

So I dunno  what is the best way to handle this: do you usually put a swicth on your unit test file to run specific test based on the box architecture and os you are running on? Let me know what you want me to do :D

```
# python3 -m unittest vdb/tests/testrecon.py

<OUTPUT TRUNCATED>
._parseSectionRelocs: Reloc section differs from Dynamics: 0xb14
_parseSectionRelocs: Reloc section differs from Dynamics: 0xb4c
Failed to resolve breakpoint address for expression: libc.exit (delayed resolution?)
/home/cmaruti/vivisect/vivtestfiles-master/linux/i386/chgrp.llvm: missing operand
Try '/home/cmaruti/vivisect/vivtestfiles-master/linux/i386/chgrp.llvm --help' for more information.
RECON: 3452 0xf7dead80 libc.exit (libc._IO_2_1_stderr_, 0x00000001, 0xf7f94884, chgrp., chgrp.calloc())
.
----------------------------------------------------------------------
Ran 2 tests in 22.185s

OK
```

**Review by atlas0fd00m — APPROVED** [review]
awesome.  well done, @cmaruti and thank you!

**Pull request merged**